src/kernels/helpers.ts

```diff
@@ -1988,17 +1991,117 @@ function convertContextToFunction(context: 'start' | 'interrupt' | 'restart') {
     }
 }
 
+const connections = new WeakMap<
+    NotebookDocument,
+    {
+        kernel: Deferred<{
+            kernel: IKernel;
+            controller: VSCodeNotebookController;
+            deadKernelAction?: 'deadKernelWasRestarted' | 'deadKernelWasNoRestarted';
+        }>;
+        options: IDisplayOptions;
+    }
+>();
+
 export async function wrapKernelMethod(
     initialController: VSCodeNotebookController,
     initialContext: 'start' | 'interrupt' | 'restart',
     serviceContainer: IServiceContainer,
     resource: Resource,
-    notebook: NotebookDocument
-) {
+    notebook: NotebookDocument,
+    options: IDisplayOptions = new DisplayOptions(false)
+): Promise<IKernel> {
+    let currentPromise = connections.get(notebook);
+    if (!options.disableUI && currentPromise?.options.disableUI) {
+        currentPromise.options.disableUI = false;
+    }
+    // If the current kernel has been disposed or in the middle of being disposed, then create another one.
+    // But do that only if we require a UI, else we can just use the current one.
+    if (
+        !options.disableUI &&
+        currentPromise?.kernel.resolved &&
+        (currentPromise?.kernel.value?.kernel?.disposed || currentPromise?.kernel.value?.kernel?.disposing)
+    ) {
+        connections.delete(notebook);
+        currentPromise = undefined;
+    }
+    const verifyKernelState = async (
```

Comment:
I feel like this would be better as function external to this method. It makes it a bit cluttered to have this as an internal function. It looks like it doesn't modify the internal state of the function, so I believe it should be moveable.